### PR TITLE
[16.0][ADD] purchase_packaging_default

### PR DESCRIPTION
**Pull request by santostelmo**
Allows you to set a default packaging on the order line using the first packaging defined in the configuration of the respective product
Default packaging quantity is then set based on purchase order line quantity divided by product packaging quantity. 
Packaging quantity is rounded up as we sell it by units. For example: We won't sell 0.5 Boxes or 1.5 Pallets.

### CODE REVIEW

File: purchase_packaging_default/models/purchase_order_line.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = "purchase.order.line"
+
+    def _get_product_packaging(self):
+        """From product get 1st packaging found ordered by sequence"""
+        product_template = self.product_id.product_tmpl_id
+        for package in product_template.packaging_ids:
+            return package
```
Comment (simahawk) [inline purchase_packaging_default/models/purchase_order_line.py]:
```suggestion
        return fields.first(product_template.packaging_ids)
```

Comment (simahawk) [reply purchase_packaging_default/models/purchase_order_line.py]: To be consistent w/ the sale module can we name this `_get_default_packaging`?

File: purchase_packaging_default/models/purchase_order_line.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            if product_packaging:
+                self.product_packaging_id = product_packaging
+
+    @api.depends("product_id", "product_qty", "product_uom")
+    def _compute_product_packaging_id(self):
+        if self.env.user.company_id.packaging_default_enabled:
+            _self = self.with_context(keep_product_packaging=True)
+            return super(PurchaseOrderLine, _self)._compute_product_packaging_id()
+        else:
+            return super(PurchaseOrderLine, self)._compute_product_packaging_id()
```
Comment (simahawk) [inline purchase_packaging_default/models/purchase_order_line.py]:
```suggestion
        _self = self
        if self.env.user.company_id.packaging_default_enabled:
            _self = _self.with_context(keep_product_packaging=True)
        return super(PurchaseOrderLine, _self)._compute_product_packaging_id()
```

File: purchase_packaging_default/models/purchase_order_line.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    @api.depends("product_packaging_id", "product_uom", "product_qty")
+    def _compute_product_packaging_qty(self):
+        """
+        Super computes product_packaging_qty rounding it to packaging_uom.precision.
```
Comment (simahawk) [inline purchase_packaging_default/models/purchase_order_line.py]: this is a comment

File: purchase_packaging_default/models/res_company.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Copyright 2023 Quartile Limited
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    packaging_default_enabled = fields.Boolean(
```
Comment (simahawk) [inline purchase_packaging_default/models/res_company.py]:
```suggestion
    purchase_packaging_default_enabled = fields.Boolean(
```

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by santostelmo** [comment]
@simahawk code updated

File: purchase_packaging_default/models/purchase_order_line.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    _inherit = "purchase.order.line"
+
+    def _get_default_packaging(self):
+        """From product get 1st packaging found ordered by sequence"""
+        product_template = self.product_id.product_tmpl_id
+        return fields.first(product_template.packaging_ids)
+
+    @api.onchange("product_id")
+    def _onchange_product_id(self):
+        if self.env.user.company_id.purchase_packaging_default_enabled:
```
Comment (simahawk) [inline purchase_packaging_default/models/purchase_order_line.py:19]:
```suggestion
        if self.env.company.purchase_packaging_default_enabled:
```
non blocking of course

**Review by simahawk — APPROVED** [review]

**Review by cyrilmanuel — APPROVED** [review]

**Comment by simahawk** [comment]
@santostelmo can you have a look at the failing test? https://github.com/OCA/purchase-workflow/actions/runs/8437473834/job/23107367001?pr=2174#step:8:966
Seems unrelated. You can open another PR to fix it.

**Comment by santostelmo** [comment]
> @santostelmo can you have a look at the failing test? https://github.com/OCA/purchase-workflow/actions/runs/8437473834/job/23107367001?pr=2174#step:8:966 Seems unrelated. You can open another PR to fix it.

I tried to replicate on local without success 😔

**Comment by simahawk** [comment]
I'll have a look

**Comment by santostelmo** [comment]
@simahawk CI is green now

**Review by cyrilmanuel — APPROVED** [review]

**Pull request merged**